[Sharding] Introduce cluster configuration

agourlay opened the pull request:
This PR introduce a proper configuration for the distributed deployment. #495

One decision I took is to separate the `consensus` configuration from the strictly `p2p` configuration as there are kind of orthogonal.

The cluster deployment is still hidden behind the compile time flag `--features consensus`, it will be addressed later on in #511.

agourlay commented:
@fabriziobonavita You might be interested in this part of the configuration, I'd be happy to receive your feedback if you have any.

agourlay commented on lib/collection/src/shard/remote_shard.rs:
```diff
@@ -39,11 +39,9 @@ pub struct RemoteShard {
     pub(crate) collection_id: CollectionId,
     pub peer_id: PeerId,
     ip_to_address: Arc<std::sync::RwLock<HashMap<u64, Uri>>>,
+    p2p_grpc_timeout: Duration,
```
I kinda cheated here as the `RemoteShard` are not yet instantiated.
The configuration will have to be injected later on.

agourlay commented on src/consensus.rs:
```diff
@@ -44,12 +42,13 @@ impl Consensus {
+        config: ConsensusConfig,
     ) -> anyhow::Result<(Self, SyncSender<Message>)> {
-        let config = Config {
+        let raft_config = Config {
```
renamed to `raft_config` to free the `config` binding for the `ConsensusConfig`

e-ivkov replied:
Maybe import the type as `RaftConfig`?

generall commented:
It does not look like users would enjoy tweaking all this parameters on their own.
So I would keep only the most essential ones:

- if distributed deployment is enabled
- Which port to use
- one single timeout parameter for all internal communication

Even I have no idea what exactly `max_in_flight_messages`, and no user should. 

Additionally, until we release the DD version, main `config.yaml` should not contain this configuration parameters, otherwise it might create confusion.

generall reviewed (changes requested):
:point_up:

agourlay commented:
@generall thanks for the comment!

> It does not look like users would enjoy tweaking all this parameters on their own.

One missing piece here is the implementation of defaults via Serde's attributes.
This way we keep the possibility to have advanced configuration knobs without confusing the users.

> one single timeout parameter for all internal communication

I am not sure about this one, the Raft messages regarding the cluster topology have a different runtime characteristic than a potentially heavy search request forwarded to a remote peer.

But I won't die on the hill for this one.

> Additionally, until we release the DD version, main config.yaml should not contain this configuration parameters, otherwise it might create confusion.

We could have a default `ClusterConfig` Serde attribute to hide this configuration for all single node users.

There is for the moment no toggle switch, maybe it would have been better to start with #511 instead :thinking:

generall commented:
> There is for the moment no toggle switch, maybe it would have been better to start with https://github.com/qdrant/qdrant/issues/511 instead thinking

:heavy_plus_sign:

e-ivkov commented:
> I am not sure about this one, the Raft messages regarding the cluster topology have a different runtime characteristic than a potentially heavy search request forwarded to a remote peer.
> But I won't die on the hill for this one.

We can start with one timeout for all of them for now I guess, then split if we notice that we need more fine-tuning there

e-ivkov commented on src/settings.rs:
```diff
@@ -49,9 +67,34 @@ pub struct Settings {
     pub cluster: ClusterConfig,
 }
 
+fn default_cluster_config() -> ClusterConfig {
+    ClusterConfig {
+        enabled: false, //disabled by default
+        p2p: P2pConfig::default(),
+        consensus: ConsensusConfig::default(),
+    }
+}
```
Maybe `impl Default for ClusterConfig` ?

e-ivkov commented on src/settings.rs:
```diff
@@ -8,17 +8,91 @@ pub struct ServiceConfig {
+
+#[derive(Debug, Deserialize, Clone)]
+pub struct ConsensusConfig {
+    #[serde(default = "default_max_in_flight_messages")]
+    pub max_in_flight_messages: usize,
```
Probably needs a comment about back-pressure?

e-ivkov replied:
Might be less confusing I think
```suggestion
    pub max_message_queue_size: usize,
```

agourlay commented:
I believe I have addressed all comments, the cluster config is now:
- absent by default from the `config.yaml`
- disabled by default
- each and every field has a default value so that users will be able to pick & choose what they need
